init: Make entry init-function less and introduce service objects

tbursztyka opened the pull request:
- optimizing ROM utilization 
- introducing de-facto a new "service" object to do so (but not exposed via any public API)

Up for a discussion. But I am still planning on removing the priority from the device driver model (and generate it automatically), following what I started here #79340, and service is going to be used also then. (I foresee that we could even have more complicated 'service' that can be stopped/restarted at run-time etc... all with device dependencies, thus being able to react on relevant PM events and so on.)

tbursztyka committed:
init: Make entry init-function less and introduce service objects

Since the addition of deinit operation in device, init and deinit have
been within each device, rendering their init entry's init function
useless.

In order to save ROM space, let's remove the init function from
init entry altogether, and introduce a new object called "service"
which owns an init function to go along with SYS_INIT/SYS_INIT_NAMED.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

andyross commented on kernel/init.c:
```diff
@@ -337,6 +340,12 @@ static int do_device_init(const struct device *dev)
+static inline bool is_entry_about_service(const void *obj)
+{
+	return (obj >= (void *)_service_list_start &&
+		obj < (void *)_service_list_end);
```
This per-item test seems needless.  You're sorting them in the linker already as a "sub-section" of an existing contiguous level.  No need to check every item, just walk the prefix that is "devices" and the suffix which is "services", no?

tbursztyka replied:
I am afraid I don't see what you are seeing. Service and device presence in init entries is shuffled, so one after another an init entry might point to a service or a device. What am I missing?

andyross reviewed (commented):
Efficiency/code-size nitpick on how the linker layout is interpreted.  But broadly no complaints, this looks reasonable.

cfriedt commented on kernel/init.c:
```diff
@@ -365,17 +374,26 @@ static void z_sys_init_run_level(enum init_level level)
-		const struct device *dev = entry->dev;
 		int result = 0;
 
+		if (unlikely(entry->_init_object == NULL)) {
```
When does this occur? I'm drawing blanks for some reason. It would be good to add a comment.

tbursztyka replied:
It's never supposed to occur actually, some prototyping left-over.

cfriedt replied:
Should it just be removed in that case? @gmarull - do you have any input here?

tbursztyka replied:
I will remove it, hopefully an update today. There is a weird bug in some builds around the macros (which does not make much sense at first glance)

cfriedt commented on include/zephyr/service.h:
```diff
@@ -0,0 +1,97 @@
+ * @defgroup service System service
+ * @ingroup os_services
+ *
+ * A system service is a one-shot initialized software part, instantiated
```
Component might sound a bit better here. Just a suggestion.

```suggestion
 * A system service is a one-shot initialized software component, instantiated
```

tbursztyka committed:
scripts: Removing init function pointer from check_init_priorities.py

As the pointer has been removed from struct init_entry, there is no point
to look for it nor to display any information about it (obviously).

Fixing the test script as well.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
tests: misc: Fixing expected output on check_init_priorities test

check_init_priorities.py does no longer output any information about any
initialization function pointer as such pointer has been removed from
struct init_entry.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
cmake: Add service section to commo-rom linker generator

struct init_entry may point to a struct device or a struct service,
each ones belonging to device or service iterable section.
Service section being newly added, handling it in cmake relevantly.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

peter-mitsis reviewed (commented):
This seems reasonable to me.

tbursztyka commented:
> This seems reasonable to me.

Would you approve then? :)
